man/man1/md2man.1
.\" Automatically generated by Pandoc 2.19.2
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "MD2MAN" "1" "December 06, 2021" "md2man 4.0" "User Manual"
.hy
.SH NAME
.PP
\f[B]md2man\f[R] - uses markdown input and pandoc to generate man pages
.SH SYNOPSIS
.PP
\f[B]md2man\f[R] [-n command-name] [-s man-section] [-v command-version]
[-aqu]
.SH DESCRIPTION
.PP
The \f[I]md2man\f[R] command generates a markdown template for input (if
none exists) with command name, sections, version, and as much info as
it can derive from command line arguments and git config.
It then generates a man page using the markdown as input.
.PP
See <https://pandoc.org/> for more info on Pandoc.
.SH COMMAND LINE OPTIONS
.TP
\f[B]-u\f[R]
displays this usage message
.TP
\f[B]-a\f[R]
indicates generate all man pages from markdown found in
\f[I]./markdown\f[R]
.TP
\f[B]-q\f[R]
indicates quiet execution, no messages or prompts
.TP
\f[B]-n command-name\f[R]
specifies the name of the command (without section)
.TP
\f[B]-s man-section\f[R]
specifies the specifies the man page section number
.TP
\f[B]-v command-version\f[R]
specifies the version of the command
.SH FILES
.PP
Markdown input and man page output are located by default in
HOME/src/doc/.
These default locations can be overridden using command line arguments.
.PP
If there is a \f[I]markdown\f[R] subdirectory in the current working
directory, then that directory is used for markdown input.
Similarly, if there is a \f[I]man\f[R] subdirectory in the current
working directory then that directory is used for man page output.
.SH EXAMPLES
.TP
\f[B]md2man -n foo -s 3 -v 2.0.1\f[R]
Generate markdown input for command name \f[I]foo\f[R] version
\f[I]2.0.1\f[R] in manual section \f[I]3\f[R].
Use the resulting markdown, \f[I]markdown/foo.3.md\f[R], as input to
generate the man page \f[I]man/man3/foo.3\f[R].
.TP
\f[B]md2man -a\f[R]
Generate man pages for all markdown files found in
\f[I]./markdown/*.md\f[R]
.SH AUTHORS
.PP
Written by Ronald Record <<github@ronrecord.com>>
.SH LICENSING
.PP
MD2MAN is distributed under an Open Source license.
See the file LICENSE in the MD2MAN source distribution for information
on terms & conditions for accessing and otherwise using MD2MAN and for a
DISCLAIMER OF ALL WARRANTIES.
.SH BUGS
.PP
Submit bug reports online at:
<<https://gitlab.com/doctorfree/DoctorFreeScripts/issues>>
.SH SEE ALSO
.PP
\f[B]man\f[R](1)
.PP
Full documentation and sources at:
<<https://gitlab.com/doctorfree/DoctorFreeScripts>>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

Jan 27, 2022 Version 4.0.1 Release 2
    Minor release update with improved support for SD management utilities,
    extending supported platforms to Linux for the SD utils.

Jan 17, 2022 Version 4.0.1 Release 1
    Release update to Version 4.0.1 including additional man pages
    and Zip format distribution archive. Minor bug fixes and improvements.
    RPM installation fixes.

Nov 26, 2021 Version 4.0 Release 1
    This release adds Debian and RPM format installation package support with
    improved support for ease of installation, configuration, update, and removal.

Wed Jan 26 14:27:43 2022 -0800 b752940 :
   Exec /bin/bash in systemd service to comply with Fedora restrictions
Tue Jan 25 08:37:12 2022 -0800 f38080d :
   Add journalclean command
Tue Jan 25 07:14:52 2022 -0800 d4b92fc :
   Set backgrounds folder to /media/user/Transcend by default on Linux
Mon Jan 24 18:21:57 2022 -0800 363a940 :
   Change default SD device to /dev/sdd
Mon Jan 24 10:17:35 2022 -0800 3be3fab :
   Add documentation for sdmount and sdumount
Sun Jan 23 19:48:26 2022 -0800 4088f8d :
   Add sdmount and sdumount commands
Sun Jan 23 10:31:20 2022 -0800 68b537d :
   Doc for sdlist
Sun Jan 23 10:04:48 2022 -0800 04f8708 :
   Add sdlist command
Sat Jan 22 16:03:14 2022 -0800 02729e7 :
   Extend SD utilities with Linux support
Fri Jan 21 07:50:47 2022 -0800 6be5d58 :
   Muriel aliases
Thu Jan 20 11:45:13 2022 -0800 24d045d :
   Redirect stderr and stdout to parent process during RPM scriptlets
Tue Jan 18 17:17:31 2022 -0800 2152b2d :
   Use DISPLAY env var if set
Tue Jan 18 13:20:58 2022 -0800 85da437 :
   Add doctorinfo script
Mon Jan 17 13:52:51 2022 -0800 65cd49e :
   Adopt release creation during Gitlab CI
Mon Jan 17 13:42:38 2022 -0800 735ff4e :
   Do not invoke dl_fred_scripts during postinstall, new version string
Mon Jan 17 12:06:57 2022 -0800 6bc24c5 :
   Move pvoutput to MirrorCommand
Mon Jan 17 12:06:31 2022 -0800 8dd49ed :
   Bump release
Fri Jan 14 13:52:28 2022 -0800 1ab3135 :
   Update miner scripts
Fri Jan 14 10:49:45 2022 -0800 988389c :
   Add disable-update-notifier and pkg/tools/get_release
Mon Jan 10 12:23:00 2022 -0800 6e4799a :
   Do not package or export gdrive utils
Mon Jan 10 12:19:41 2022 -0800 9c85af0 :
   Remove gdrive files as they are now packaged in their own repo
Sun Jan 9 17:43:49 2022 -0800 8962417 :
   disable rpmbuild strip
Sun Jan 9 17:18:32 2022 -0800 5af1d70 :
   Fix files section of spec file
Sun Jan 9 17:10:20 2022 -0800 905d402 :
   Move all files into /usr/local
Sun Jan 9 16:39:42 2022 -0800 e27d740 :
   Try noninteractive rpm install again
Sun Jan 9 15:35:04 2022 -0800 64ddc3a :
   Fix for CI
Sun Jan 9 15:28:21 2022 -0800 fd5915b :
   export noninteractive
Sun Jan 9 15:18:00 2022 -0800 c6e51e9 :
   Copy rpm dist tree in install phase
Sun Jan 9 14:46:58 2022 -0800 bfa4e07 :
   Install rpm in CI in noninteractive mode with TZ set
Sun Jan 9 14:17:31 2022 -0800 53255f8 :
   Use ubuntu:latest CI image
Sun Jan 9 14:09:05 2022 -0800 c6963e9 :
   Version 4.0 release 3
Sun Jan 9 14:04:52 2022 -0800 fd33ea1 :
   Install zip if not found, create zip artifacts in CI
Thu Jan 6 11:57:39 2022 -0800 3b3f004 :
   Add drive to projects
Mon Jan 3 20:15:25 2022 -0800 748b3e0 :
   Add author attribution
Mon Jan 3 16:22:13 2022 -0800 236f864 :
   Add -l option to list sync folders
Mon Jan 3 16:11:26 2022 -0800 d30988a :
   Renamed getfileids as getfolderids
Mon Jan 3 13:01:47 2022 -0800 97d0e8d :
   Check for gdrive binary in PATH
Mon Jan 3 12:38:50 2022 -0800 ded8f3d :
   Add options to getfolderids, start on support for spaces in file and folder names
Mon Jan 3 10:24:36 2022 -0800 627a90d :
   Support spaces in folder names, retrieve all folder ids
Sun Jan 2 21:10:06 2022 -0800 1ec7cb9 :
   Export gdrive utils
Sun Jan 2 21:07:06 2022 -0800 40e2bdb :
   Add gdrive scripts to exports
Sun Jan 2 21:04:30 2022 -0800 e9d2a92 :
   Add comment for developer's guide to api
Sun Jan 2 20:19:56 2022 -0800 c674bc0 :
   Add sync_from_drive, update sync2drive
Sun Jan 2 11:48:46 2022 -0800 285acd8 :
   Updated gdrive utility scripts
Sat Jan 1 20:14:40 2022 -0800 a670a0d :
   Add new scripts for command line management of Google Drive
Fri Dec 31 08:52:06 2021 -0800 a1659d2 :
   Add Mirco Cabbia alias
Tue Dec 28 09:50:38 2021 -0800 904ab6d :
   Rename MirrorCommandLine to MirrorCommand
Mon Dec 27 18:20:34 2021 -0800 3554d70 :
   New model alias
Mon Dec 27 08:43:35 2021 -0800 5a2a524 :
   New model alias
Mon Dec 27 08:23:33 2021 -0800 a981875 :
   Create man page for md2man
Sun Dec 26 17:48:24 2021 -0800 6d83be7 :
   Set output dir based on man page section in md2man
Sat Dec 25 08:04:35 2021 -0800 ef79abb :
   Added Colombian wallpapers
Fri Dec 24 11:34:02 2021 -0800 0c9e9cc :
   Adjusted aliases
Thu Dec 23 09:38:25 2021 -0800 c2b5191 :
   Updated model name, add Mila Azul to Ukrainians
Wed Dec 22 16:31:35 2021 -0800 d4296b4 :
   Use HEAD to determine branch
Tue Dec 21 14:01:20 2021 -0800 526124a :
   Bump release number, update changelog
Tue Dec 21 10:26:02 2021 -0800 8c3a636 :
   Add all my clone and build scripts
Tue Dec 21 09:43:04 2021 -0800 d625a49 :
   Check and set location of Synergy client binary and system tools in Synergy startup control script
Tue Dec 21 09:14:43 2021 -0800 7a101e1 :
   Add man page for sdwrite, add SEE ALSO pages to all SD card commands
Tue Dec 21 08:40:29 2021 -0800 58b2dfc :
   Move sdwrite from MirrorCommandLine package to here
Sat Dec 18 17:18:51 2021 -0800 148f60a :
   Set synergyctl IP in postinstall
Sat Dec 18 16:39:57 2021 -0800 a8df6ce :
   Add synergyctl script to control Synergy start/stop
Fri Dec 17 07:15:45 2021 -0800 30fd5e5 :
   New model aliases
Mon Dec 13 21:11:37 2021 -0800 2c51ae3 :
   Fix typo in rmbrokenpkg
Mon Dec 13 20:18:02 2021 -0800 e5436b5 :
   Add rmbrokenpkg
Sun Dec 12 13:40:26 2021 -0800 2a4714d :
   Add note about Windows System for Linux

scripts/md2man.sh
#!/bin/bash
#
# md2man - Uses Github flavored markdown input and pandoc to generate man pages
#    md2man generates a markdown template for input (if none exists)
#    with sections and as much info as it can derive from command line arguments.
#
#    See https://pandoc.org/ for Pandoc info
#
# Author: Ron Record <gitlab@ronrecord.com>
#

# Default input/output directory
MANDIR=${HOME}/src/doc
OUTDIR=${MANDIR}
INPDIR=${MANDIR}
ALL=
USAGE=
QUIET=
HERE=`pwd`
DATE=`date +'%B %d, %Y'`

# Are we in a git repository with remote origin url?
# If so then this should return something of the form
# ssh://gitlab.com/doctorfree/DoctorFreeScripts.git
URL=`git config --get remote.origin.url`
if [ "${URL}" ]
then
  GIT_HOST=`echo ${URL} | awk -F '/' ' { print $3 } '`
  GIT_USER=`echo ${URL} | awk -F '/' ' { print $4 } '`
  PROJECT=`echo ${URL} | awk -F '/' ' { print $5 } ' | sed -e "s/\.git//"`
else
  GIT_HOST="GIT_HOST"
  GIT_USER="GIT_USERNAME"
  PROJECT="PROJECT"
fi

NAME=`git config --get user.name`
EMAIL=`git config --get user.email`
[ "${NAME}" ] || NAME="YOUR NAME"
[ "${EMAIL}" ] || EMAIL="YOUR EMAIL"

# Default command name, version, and man section
comname=""
VERSION="1.0.0"
section="1"

usage() {
    printf "\nUsage:\n\tmd2man [-n command-name] [-s man-section] "
    printf "[-v command-version] [-aqu]"
    printf "\n\tWhere:\n\t\t-u displays this usage message"
    printf "\n\t\t-a indicates generate all man pages from markdown found in ./markdown"
    printf "\n\t\t-q indicates quiet execution, no messages or prompts"
    printf "\n\t\t-n command-name specifies the name of the command (without section)"
    printf "\n\t\t-s man-section specifies the specifies the man page section number"
    printf "\n\t\t-v command-version specifies the version of the command"
    printf "\n\tMarkdown input located in ${INPDIR}"
    printf "\n\tMan page output located in ${OUTDIR}"
    printf "\n\tAuthor Name and Email are retrieved from git global config variables"
    printf "\n\t(Currently using Name=${NAME} and Email=${EMAIL})"
    printf "\nSummary:\n\tmd2man uses markdown input and pandoc"
    printf " to generate man pages"
    printf "\nExample:\n\tmd2man -n foo -s 3 -v 2.0.1"
    printf "\nDescription:\n\tGenerates a markdown template for input (if none exists)"
    printf "\n\twith command name, sections, version, and as much info"
    printf "\n\tas it can derive from command line arguments and git config."
    printf "\n\tSee https://pandoc.org/ for Pandoc info."
    printf "\nAuthor:\n\t${NAME} <${EMAIL}>\n\n"
    exit 1
}

mkmanfrom() {
    command_name="$1"
    command_sect="$2"
    OUTDIR=${MANDIR}/man${command_sect}
    [ -d "${OUTDIR}" ] || mkdir -p "${OUTDIR}"
    # Use awk to get Unicode support
    capcomm=`echo "${command_name}" | awk 'BEGIN { getline; print toupper($0) }'`

    [ -f "${INPDIR}/${command_name}.${command_sect}.md" ] || {
        echo "---
title: ${capcomm}
section: ${command_sect}
header: User Manual
footer: ${command_name} ${VERSION}
date: ${DATE}
---
# NAME
**${command_name}** - PLACE BRIEF SUMMARY HERE

# SYNOPSIS
**${command_name}** [ **OPTION** ARG ]

# DESCRIPTION
PLACE DESCRIPTION HERE

# COMMAND LINE OPTIONS
**-u**
: display usage message

**-n**
: ADDITIONAL OPTIONS

# EXAMPLES
**${command_name} -n Arg**
: EXAMPLE DESCRIPTION GOES HERE

# AUTHORS
Written by ${NAME} &lt;${EMAIL}&gt;

# LICENSING
${capcomm} is distributed under an Open Source license.
See the file "LICENSE" in the ${capcomm} source distribution
for information on terms &amp; conditions for accessing and
otherwise using ${capcomm} and for a DISCLAIMER OF ALL WARRANTIES.

# BUGS
Submit bug reports online at: &lt;https://${GIT_HOST}/${GIT_USER}/${PROJECT}/issues&gt;

# SEE ALSO
Full documentation and sources at: &lt;https://${GIT_HOST}/${GIT_USER}/${PROJECT}&gt;
" > "${INPDIR}/${command_name}.${command_sect}.md"
    }

    [ "${QUIET}" ] || {
        echo "Using ${INPDIR}/${command_name}.${command_sect}.md"
        echo "as markdown input"
        echo ""
        echo "Using ${OUTDIR}/${command_name}.${command_sect}"
        echo "as man page output"
        echo ""
    }

    # Use Github flavored markdown as input, Man page format as output
    pandoc -f gfm+definition_lists -s -t man \
           -o ${OUTDIR}/${command_name}.${command_sect} \
              ${INPDIR}/${command_name}.${command_sect}.md 

    [ "${QUIET}" ] || {
      while true
      do
        read -p "Do you wish to view the generated ${command_name}.${command_sect} man page ? ('Y'/'N'): " yn
        case $yn in
          [Yy]*)
              if [ "$plat" == "Darwin" ]
              then
                  man -M `dirname ${OUTDIR}` ${command_name}
              else
                  man -l ${OUTDIR}/${command_name}.${command_sect}
              fi
              break
              ;;
          [Nn]*)
              break
              ;;
          * )
              echo "Please answer yes or no."
              ;;
        esac
      done
    }
}

[ -f VERSION ] && . ./VERSION

while getopts an:s:v:qu flag; do
    case $flag in
        a)
            ALL=1
            QUIET=1
            ;;
        n)
            comname="$OPTARG"
            ;;
        s)
            section="$OPTARG"
            ;;
        v)
            VERSION="$OPTARG"
            ;;
        q)
            QUIET=1
            ;;
        u)
            USAGE=1
            ;;
	esac
done
shift $(( OPTIND - 1 ))

# Check for markdown and man subdirs in current directory
# If both directories exist then set MANDIR and OUTPUT dirs
[ -d markdown ] && [ -d man ] && {
    INPDIR=${HERE}/markdown
    MANDIR=${HERE}/man
    OUTDIR=${MANDIR}/man${section}
}

[ "${USAGE}" ] && usage

[ -d "${INPDIR}" ] || mkdir -p "${INPDIR}"
[ -d "${OUTDIR}" ] || mkdir -p "${OUTDIR}"

inst_pandoc=`type -p pandoc`
plat=`uname -s`

[ "${inst_pandoc}" ] || {
    if [ "$plat" == "Darwin" ]
    then
      INSTCOMM="brew install pandoc"
    else
      debian=
      [ "${ID_LIKE}" == "debian" ] && debian=1
      [ "${debian}" ] || [ -f /etc/debian_version ] && debian=1
      if [ "${debian}" ]
      then
          INSTCOMM="sudo apt install pandoc"
      else
          INSTCOMM="sudo yum install pandoc"
      fi
    fi
    echo "Cannot locate pandoc in the execution PATH."
    echo "Install pandoc with the command:"
    printf "\n\t${INSTCOMM}\n\n"
    while true
    do
      read -p "Do you wish to install pandoc now ? ('Y'/'N'): " yn
      case $yn in
          [Yy]*)
              ${INSTCOMM}
              break
              ;;
          [Nn]*)
              echo "Exiting."
              exit 1
              break
              ;;
          * )
              echo "Please answer yes or no."
              ;;
      esac
    done
}

[ "${ALL}" ] && {
  # Check to see if there are any markdown files to process
  for mdf in markdown/*.md
  do
    [ "${mdf}" == "markdown/*.md" ] && ALL=
  done
}

if [ "${ALL}" ]
then
  for mdf in markdown/*.md
  do
    [ "${mdf}" == "markdown/*.md" ] && continue
    [ "${mdf}" == "markdown/README.md" ] && continue
    comname=`basename ${mdf} | awk -F '.' ' { print $1 } '`
    comsect=`basename ${mdf} | awk -F '.' ' { print $2 } '`
    echo "Processing ${comname}.${comsect}"
    mkmanfrom ${comname} ${comsect}
  done
else
  [ "${comname}" ] || {
    if [ "$1" ]
    then
      comname=$*
    else
      echo "No command name provided. Using 'COMMAND' as command name."
      comname="COMMAND"
    fi
  }
  for mdf in ${comname}
  do
    mkmanfrom ${mdf} ${section}
  done
fi
